rclcpp/src/rclcpp/experimental/executors/events_executor/events_executor.cpp

```diff
@@ -386,11 +386,15 @@ EventsExecutor::get_automatically_added_callback_groups_from_nodes()
 void
 EventsExecutor::refresh_current_collection_from_callback_groups()
 {
+  // Build the new collection
   this->entities_collector_->update_collections();
   auto callback_groups = this->entities_collector_->get_all_callback_groups();
   rclcpp::executors::ExecutorEntitiesCollection new_collection;
   rclcpp::executors::build_entities_collection(callback_groups, new_collection);
 
+  // Acquire lock before modifying the current collection
+  std::lock_guard<std::recursive_mutex> lock(collection_mutex_);
```

Comment:
this needs to protect `add_notify_waitable_to_collection(new_collection.waitables)`? if not, it would be probably nice to minimize the lock window to move this mutex lock right before `add_notify_waitable_to_collection(current_entities_collection_->waitables)`?

Reply:
Yes, you are right.
Moved this to minimize the lock